[Docs] ES|QL Data Federation (#146945)
elastic/elasticsearch: docs/reference/query-languages/esql/esql-data-federation-security.md

---
navigation_title: "Manage access"
description: "Credential encryption, credential masking, and privileges for ES|QL federated data sources and datasets."
applies_to:
  stack: preview =9.5
  serverless: preview
products:
  - id: elasticsearch
---

# Manage credentials and privileges for {{esql}} Data Federation

Because federated data lives outside {{es}}, connecting to a private store means {{es}} holds that data source's credentials and must control who can manage connections and read external data. This page covers credential encryption and masking, and the privilege model for data sources and datasets.

## Credentials

When a data source connects to a private bucket, its credentials are stored in the cluster state. {{es}} protects these credentials with encryption at rest and redaction in API responses.

### Credential encryption

<!-- TODO: restore the link to the project encryption key page once PR #152731 (docs/reference/elasticsearch/project-encryption-key.md) merges. Linking now breaks the build because the target does not exist on this branch. Link markup: [project encryption key](../../elasticsearch/project-encryption-key.md) -->

> When a data source includes credentials, {{es}} encrypts them before writing them to the cluster state. It uses the cluster's project encryption key, a single key that {{es}} manages on your behalf. Federated data is the first feature to rely on this key.

Reviewer: ```suggestion
When a data source includes credentials, {{es}} encrypts them using cluster state encryption.
```

It needs a link to the cluster encryption key page.

Author: yep there's a TODO comment above to track this

The cluster state encryption key is available automatically in most environments, including {{ech}}, {{ece}}, {{eck}}, and {{serverless-short}}.

> By default, if the cluster state encryption key is not available when you create a data source, a `PUT /_query/data_source` request that includes credentials returns a `503` error. On upgraded self-managed clusters, the encryption password is not configured automatically. Add `cluster.state.encryption.password.<id>` and `cluster.state.encryption.active_password_id` to the keystore on every node, then call `POST /_nodes/reload_secure_settings`.

Reviewer 2: The encryption opt-out stores secrets unencrypted — worth a caution given this page's emphasis.

```suggestion
By default, if the cluster state encryption key is not available when you create a data source, a `PUT /_query/data_source` request that includes credentials returns a `503` error. On upgraded self-managed clusters, the encryption password is not configured automatically. Add `cluster.state.encryption.password.<id>` and `cluster.state.encryption.active_password_id` to the keystore on every node, then call `POST /_nodes/reload_secure_settings`.

:::{warning}
Setting `cluster.state.encryption.required: false` lets a data source with credentials be created even when no encryption key is available. The credentials are then stored unencrypted in the cluster state, with only a warning in the logs. Keep the default unless you accept that risk.
:::
```

* On {{ech}}, {{ece}}, {{eck}}, and {{serverless-short}}, the platform provides the key's password for you.

> * On self-managed deployments, the password is generated automatically the first time a node starts with security auto-configuration, the same process that configures TLS. You configure it yourself only if security auto-configuration did not run.

Reviewer: ```suggestion
The cluster state key is available automatically in most environments, including {{ech}}, {{ece}}, {{eck}}, and {{serverless-short}}.
```

Author: 👍 just amended your comment to "cluster state key"

<!-- TODO: restore this line once PR #152731 (docs/reference/elasticsearch/project-encryption-key.md) merges. It links to a page not yet on this branch and would break the build.
For how to set the password, rotate the key, and check its health, refer to [project encryption key](../../elasticsearch/project-encryption-key.md). -->

For how to set the password, rotate the key, and check its health, refer to your deployment's encryption documentation.

> By default, if the project encryption key is not available when you create a data source, a `PUT /_query/data_source` request that includes credentials returns a `503` error. Retry once the key is available, or configure the key first.

Reviewer: ```suggestion
By default, if the cluster state encryption key is not available when you create a data source, a `PUT /_query/data_source` request that includes credentials returns a `503` error. Retry once the key is available, or configure the key first.
```

## Privileges

Dataset operations are authorized by the standard {{es}} [index privileges](../../elasticsearch/security-privileges.md#privileges-list-indices), so a role that already administers or reads the matching index names covers datasets with no additional grant.

| Operation | Privilege | Type |
|---|---|---|
| Query a dataset | `read` | Index, on the dataset name |
| Create, read, or delete a dataset | `manage` or `all` | Index, on the dataset name |
| Dataset administration granted on its own | `create_dataset`, `read_dataset_metadata`, `delete_dataset`, `manage_dataset` | Index, on the dataset name |
| Create or replace a data source | `global.data_source` `create` / `cluster.manage` | Global (fine-grained) / Cluster |
| Read a data source definition | `global.data_source` `read_metadata` / `cluster.manage` | Global (fine-grained) / Cluster |

> | Delete a data source | `global.data_source` `delete` / `cluster.manage` | Global (fine-grained) / Cluster |

Reviewer 2: Add the fifth `global.data_source` token, `read` — the least-privilege grant for attaching a dataset to a data source (the prose below already implies it).

```suggestion
| Delete a data source | `global.data_source` `delete` / `cluster.manage` | Global (fine-grained) / Cluster |
| Reference a data source from a dataset | `global.data_source` `read` / `cluster.manage` | Global (fine-grained) / Cluster |
```

| All data source operations | `global.data_source` `manage` / `cluster.manage` | Global (fine-grained) / Cluster |

Creating a dataset that references a data source also requires the `read` data source privilege for that data source. The two are authorized independently.

> The `read` privilege granted on a dataset name must not carry document-level or field-level security. `FROM <dataset>` is rejected if it does.

Reviewer 2: The same DLS/FLS restriction covers views.

```suggestion
The `read` privilege granted on a dataset name must not carry document-level or field-level security. `FROM <dataset>` is rejected if it does. The same restriction applies to [{{esql}} views](esql-views.md).
```

> `superuser` has full access to data sources and datasets. Data source management is reached only through `superuser` or a role explicitly granted `global.data_source`.

Reviewer 2: The "only superuser or `global.data_source`" line contradicts the table above and the code — cluster `manage`/`all` reaches these actions too.

```suggestion
`superuser` has full access to data sources and datasets. Data source management is reached through the cluster `manage` (or `all`) privilege, or through a role explicitly granted `global.data_source` for fine-grained control.
```

Creating a dataset that references a data source also requires the `read` data source privilege for that data source; the two are authorized independently.

The `read` privilege granted on a dataset name must not carry document-level or field-level security; `FROM <dataset>` is rejected if it does.

`superuser` has full access to data sources and datasets. Data source management is reached only through `superuser` or a role explicitly granted `global.data_source`.

> A role configures these privileges as follows. The example grants querying `sales` and `clicks`, dataset administration over `acme_*`, and management of the `acme_*` data sources:

Reviewer: This is an advanced configuration, we should at least call it out explicitly and maybe add the more generic one using `manage` cluster privilege.

Author: Addressed — the table now shows both paths (fine-grained `global.data_source` and generic `cluster.manage`), and the prose below calls out `manage`/`all` as the simpler option.

## Credentials

When a data source connects to a private bucket, its encrypted credentials are stored in the cluster state. {{es}} protects these credentials with encryption at rest and redaction in API responses.

### Credential encryption

<!-- TODO: restore the link to the cluster state encryption key page once PR #152731 (docs/reference/elasticsearch/project-encryption-key.md) merges. Linking now breaks the build because the target does not exist on this branch. Link markup: [cluster state encryption key](../../elasticsearch/project-encryption-key.md) -->
When a data source includes credentials, {{es}} encrypts them using the cluster state encryption key before storing them in the cluster state.

The cluster state encryption key is available automatically in most environments, including {{ech}}, {{ece}}, {{eck}}, and {{serverless-short}}.

<!-- TODO: restore this line once PR #152731 (docs/reference/elasticsearch/project-encryption-key.md) merges. It links to a page not yet on this branch and would break the build.
For how to set the password, rotate the key, and check its health, refer to [cluster state encryption key](../../elasticsearch/project-encryption-key.md). -->

> For how to set the password, rotate the key, and check its health, refer to your deployment's encryption documentation.

Reviewer 3: should we link to the more generic [Security page here](https://www.elastic.co/docs/deploy-manage/security/secure-your-cluster-deployment), at least temporarily (as per your note)

Author: Resolved in https://github.com/elastic/elasticsearch/commit/48b0b73a4a5d1d2a5af7a829bbb3963c0582f9b2

---
navigation_title: "Manage access"
description: "Secure ES|QL Data Federation by controlling access to data sources and datasets, encrypting credentials, and configuring privileges."
applies_to:
  stack: preview =9.5
  serverless: preview
products:
  - id: elasticsearch
---

# Manage credentials and privileges for {{esql}} Data Federation

> Because federated data lives outside {{es}}, connecting to a private store means {{es}} holds that data source's credentials and must control who can manage connections and read external data. This page covers credential encryption and masking, and the privilege model for data sources and datasets.

Reviewer 4: imo these are two different things that belong on two different pages

Author: To be followed up in https://github.com/elastic/docs-content-internal/issues/1525

## Credentials

When a data source connects to a private bucket, its encrypted credentials are stored in the cluster state. {{es}} protects these credentials with encryption at rest and redaction in API responses.

### Credential encryption

> When a data source includes credentials, {{es}} encrypts them using the cluster state encryption key before storing them in the cluster state. For general information about securing your deployment, refer to [secure your cluster or deployment](docs-content://deploy-manage/security/secure-your-cluster-deployment.md).

Reviewer 4: hm this link doesn't feel helpful here

Author: Fixed in https://github.com/elastic/elasticsearch/pull/146945/commits/affeef8e8fee

The cluster state encryption key is available automatically in most environments, including {{ech}}, {{ece}}, {{eck}}, and {{serverless-short}}.

> By default, if the cluster state encryption key is not available when you create a data source, a `PUT /_query/data_source` request that includes credentials returns a `503` error. On upgraded self-managed clusters, the encryption password is not configured automatically. Add `cluster.state.encryption.password.<id>` and `cluster.state.encryption.active_password_id` to the keystore on every node, then call `POST /_nodes/reload_secure_settings`.

Reviewer 4: what does the `<id>` placeholder mean

```suggestion
By default, if the cluster state encryption key is not available when you create a data source, a `PUT /_query/data_source` request that includes credentials returns a `503` error. 

On upgraded self-managed clusters, the encryption password is not configured automatically. To configure a cluster state encryption key, add `cluster.state.encryption.password.<id>` and `cluster.state.encryption.active_password_id` to the keystore on every node, then call `POST /_nodes/reload_secure_settings`.
```

Author: Fixed in https://github.com/elastic/elasticsearch/pull/146945/commits/118f93d4f9a0
